wgpu/backends/rs.py

```diff
@@ -176,6 +176,20 @@ def check_struct(struct_name, d):
 libf = SafeLibCalls(lib, error_handler)
 
 
+def enumerate_adapters():
+    """Return a list of all available adapters."""
+    # The first call is to get the number of adapters, and
+    # the second call is to get the actual adapters
+
+    # H: size_t f(WGPUInstance instance, WGPUInstanceEnumerateAdapterOptions const * options, WGPUAdapter * adapters)
+    adapter_count = libf.wgpuInstanceEnumerateAdapters(
+        get_wgpu_instance(), ffi.NULL, ffi.NULL
+    )
+    adapters = ffi.new("WGPUAdapter[]", adapter_count)
+    libf.wgpuInstanceEnumerateAdapters(get_wgpu_instance(), ffi.NULL, adapters)
+    return [GPUAdapter(adapter) for adapter in adapters]
```

Comment:
The codegen has trouble with the refactoring of the adapter metadata. A relatively easy way to work around that is to split that part out of `request_adapter()`, into a a new `GPU._create_adapter(adapter_id)`.

Then this here becomes:
```suggestion
    gpu = GPU()
    return [gpu._create_adapter(adapter) for adapter in adapters]
```

Reply:
When I turned _create_adapter into a member function of the GPU, a TypeError occurred in enumerate_adapters as below. Therefore, I made it a free function.

```sh
_______________________________ test_rs_enumerate_adapters ________________________________

    @mark.skipif(not can_use_wgpu_lib, reason="Needs wgpu lib")
    def test_rs_enumerate_adapters():
        # Get all available adapters
>       adapters = wgpu.backends.rs.enumerate_adapters()

tests/test_rs_basics.py:152: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def enumerate_adapters():
        """Return a list of all available adapters."""
        # The first call is to get the number of adapters, and
        # the second call is to get the actual adapters.
        # Note that the second arg (now NULL) can be a
        # `WGPUInstanceEnumerateAdapterOptions` to filter by backend.
    
        # H: size_t f(WGPUInstance instance, WGPUInstanceEnumerateAdapterOptions const * options, WGPUAdapter * adapters)
        adapter_count = libf.wgpuInstanceEnumerateAdapters(
            get_wgpu_instance(), ffi.NULL, ffi.NULL
        )
        adapters = ffi.new("WGPUAdapter[]", adapter_count)
        # H: size_t f(WGPUInstance instance, WGPUInstanceEnumerateAdapterOptions const * options, WGPUAdapter * adapters)
        libf.wgpuInstanceEnumerateAdapters(get_wgpu_instance(), ffi.NULL, adapters)
>       gpu = GPU()
E       TypeError: 'NoneType' object is not callable

wgpu/backends/rs.py:194: TypeError
```

Reply:
Ah yes, sorry. This is due to a bug in the `_register_backend` decorator, that will be fixed in #408.